prql/src/semantic/analyzer.rs

```diff
@@ -0,0 +1,276 @@
+            last: Scope::new_open(),
+            spill: Scope::default(),
+        }
+    }
+
+    pub fn get_ast(&self) -> &Node {
+        &self.ast
+    }
+
+    pub fn get_table_columns(&self) -> Vec<Option<String>> {
+        self.last
+            .table
+            .iter()
+            .filter_map(|id| self.declarations[*id].0.as_variable().cloned())
+            .map(|c| c.name)
+            .collect()
+    }
+
+    /// Runs semantic analysis on the query, using current state.
+    /// Appends query to current query.
+    pub fn append(&mut self, query: Query) -> Result<()> {
+        let mut nodes = self.fold_nodes(query.nodes)?;
+
+        if let Some(ast) = self.ast.item.as_query_mut() {
+            ast.nodes.append(&mut nodes);
+        }
+        Ok(())
+    }
+
+    fn declare(&mut self, dec: Declaration, span: Option<Span>) -> usize {
+        self.declarations.push((dec, span));
+        self.declarations.len() - 1
+    }
+
+    fn declare_func(&mut self, func_def: FuncDef) -> usize {
+        let name = func_def.name.clone();
+        let is_variable = func_def.named_params.is_empty() && func_def.positional_params.is_empty();
+
+        let span = Some(func_def.body.span);
+        let id = self.declare(Declaration::Function(func_def), span);
+
+        if is_variable {
+            self.last
+                .declare_variable(Some(&format!("$.{name}")), id, false);
+        } else {
+            self.last.functions.insert(name, id);
+        }
+
+        id
+    }
+
+    fn declare_variable(&mut self, var_dec: VarDec, span: Span) -> usize {
+        let name = var_dec.name.clone();
+
+        let id = self.declare(Declaration::Variable(var_dec), Some(span));
+
+        self.last.declare_variable(name.as_deref(), id, true);
+
+        id
+    }
+
+    fn lookup_variable(&mut self, ident: &str) -> Option<usize> {
+        let id = self.last.lookup_variable(ident);
+        if id.is_none() {
+            self.spill.declare_variable(Some(ident), 0, false);
+        }
+        id
+    }
+
+    fn declare_table_column(&mut self, position: usize, node: &Node) {
+        let position = Some(position);
+        let var_dec = if let Some(named_expr) = node.item.as_named_expr() {
+            VarDec {
+                position,
+                declaration: named_expr.expr.clone(),
+                name: Some(named_expr.name.clone()),
+            }
+        } else {
+            VarDec {
+                position,
+                declaration: Box::from(node.clone()),
+                name: None,
+            }
+        };
+        self.declare_variable(var_dec, node.span);
+    }
+}
+
+impl AstFold for SemanticAnalyzer {
+    // save functions declarations
+    fn fold_nodes(&mut self, items: Vec<Node>) -> Result<Vec<Node>> {
+        // We cut out function def, so we need to run it
+        // here rather than in `fold_func_def`.
+        let mut r = Vec::with_capacity(items.len());
+
+        for item in items {
```

Comment:
FYI we can also structure as an Iterator & `.filter_map` like below in `fold_pipeline` but same result

Reply:
I prefer for loops instead of iterators because they seem "native" and behave better with `?`. But yeah, here itearator would be more readable.

Reply:
nvm, you can do:
```
items
            .into_iter()
            .map(|item| {
                Ok(match item {
                    ... => {
                        ...
                        None
                    }
                    _ => Some(self.fold_node(item)?),
                })
            })
            .filter_map(|x| x.transpose())
            .try_collect()
```
which is better. Map has to return `Result` (to allow `?`) so it cannot be `filter_map`, but then you can just transpose `Option` and `Result`.